pkg/localstorev2/reserve.go

```diff
@@ -0,0 +1,496 @@
+
+	return &putterSession{
+		Putter: storage.PutterFunc(func(ctx context.Context, chunk swarm.Chunk) error {
+			newChunk, err := db.reserve.Put(ctx, trx, chunk)
+			if err != nil {
+				return err
+			}
+			triggerBins[db.po(chunk.Address())] = true
+			if newChunk {
+				count++
+			}
+			return nil
+		}),
+		done: func(swarm.Address) error {
+			defer db.lock.Unlock(reserveLock)
+			err := commit()
+			if err != nil {
+				return err
+			}
+			db.reserve.AddSize(count)
+			for bin := range triggerBins {
+				db.reserveBinEvents.Trigger(string(bin))
+			}
+			if !db.reserve.IsWithinCapacity() {
+				db.events.Trigger(reserveOverCapacity)
+			}
+			return nil
+		},
+		cleanup: func() error {
+			defer db.lock.Unlock(reserveLock)
+			return rollback()
+		},
+	}
+}
+
+// EvictBatch evicts all chunks belonging to a batch from the reserve.
+func (db *DB) EvictBatch(ctx context.Context, batchID []byte) error {
+	return db.evictBatch(ctx, batchID, swarm.MaxBins)
+}
+
+func (db *DB) evictBatch(ctx context.Context, batchID []byte, upToBin uint8) error {
+
+	db.lock.Lock(reserveLock)
+	defer db.lock.Unlock(reserveLock)
+
+	for b := uint8(0); b < upToBin; b++ {
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		txnRepo, commit, rollback := db.repo.NewTx(ctx)
+
+		evicted, err := db.reserve.EvictBatchBin(txnRepo, b, batchID)
+		if err != nil {
+			_ = rollback()
+			return err
+		}
+
+		err = commit()
+		if err != nil {
+			return err
+		}
+
+		db.logger.Info("reserve eviction", "bin", b, "evicted", evicted, "batchID", hex.EncodeToString(batchID), "size", db.reserve.Size())
+
+		db.reserve.AddSize(-evicted)
+	}
+
+	return nil
+}
+
+func (db *DB) unreserve(ctx context.Context) error {
+
+	radius := db.reserve.Radius()
+	defer db.events.Trigger(reserveUnreserved)
+
+	for radius < swarm.MaxPO {
+
+		err := db.batchstore.Iterate(func(b *postage.Batch) (bool, error) {
+
+			if db.reserve.IsWithinCapacity() {
+				return true, nil
+			}
+
+			err := db.evictBatch(ctx, b.ID, radius)
+			if err != nil {
+				return false, err
+			}
+
+			return false, nil
+		})
+		if err != nil {
+			return err
+		}
+		if db.reserve.IsWithinCapacity() {
+			return nil
+		}
+
+		radius++
+		db.logger.Info("reserve radius increase", "radius", radius)
+		_ = db.reserve.SetRadius(db.repo.IndexStore(), radius)
+	}
+
+	return errMaxRadius
+}
+
+// ReserveLastBinIDs returns all of the highest binIDs from all the bins in the reserve.
+func (db *DB) ReserveLastBinIDs() ([]uint64, error) {
+	return db.reserve.LastBinIDs(db.repo.IndexStore())
+}
+
+// BinC is the result returned from the SubscribeBin channel that contains the chunk address and the binID
+type BinC struct {
+	Address swarm.Address
+	BinID   uint64
+}
+
+// SubscribeBin returns a channel that feeds all the chunks in the reserve from a certain bin between a start and end binIDs.
+func (db *DB) SubscribeBin(ctx context.Context, bin uint8, start, end uint64) (<-chan *BinC, func(), <-chan error) {
```

Comment:
Would it be better to have the `errc` inside the `BinC`?

I think this would make the UX easier for the client as he would potentially need just 1 for loop.